File: apps/api/app/terminals/check.py
"""Verificação de respostas — espelho de packages/sim/src/terminals/check.ts.

As normalizações replicam a semântica do JavaScript (trim/regex) para que cliente e
servidor cheguem sempre ao mesmo veredito.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

# espaços em branco reconhecidos por String.prototype.trim e /\s/ no JavaScript
JS_WS = "\t\n\v\f\r                  　﻿"
_WS_RUN = re.compile(f"[{re.escape(JS_WS)}]+")
_DIGITS_1_3 = re.compile(r"[0-9]{1,3}\Z")
_DIGITS_1_10 = re.compile(r"[0-9]{1,10}\Z")


@dataclass(frozen=True)
class CheckResult:
    correct: bool
    items: list[bool]

    def as_dict(self) -> dict[str, Any]:
        return {"correct": self.correct, "items": list(self.items)}


def js_trim(s: str) -> str:
    return s.strip(JS_WS)


def normalize_ipv4(s: str) -> str | None:
    parts = js_trim(s).split(".")
    if len(parts) != 4:
        return None
    nums: list[int] = []
    for p in parts:
        if not _DIGITS_1_3.match(p):
            return None
        n = int(p)
        if n > 255:
            return None
        nums.append(n)
    return ".".join(str(n) for n in nums)


def normalize_int(s: str) -> str | None:
    t = js_trim(s)
    if t.startswith("+"):
        t = t[1:]
    if not _DIGITS_1_10.match(t):
        return None
    return str(int(t))


def normalize_text(s: str) -> str:
    return _WS_RUN.sub(" ", js_trim(s).lower())


def is_int(v: Any) -> bool:
    """Number.isInteger: booleanos não são números em JS."""
    if isinstance(v, bool):
        return False
    if isinstance(v, int):
        return True
    return isinstance(v, float) and float(v).is_integer()


def _is_int_list(v: Any) -> bool:
    return isinstance(v, list) and all(is_int(x) for x in v)


def effective_allowed(ports: list[int], rules: dict[str, Any]) -> list[int]:
    allowed: list[int] = []
    for port in ports:
        rule = next((r for r in rules.get("rules", []) if isinstance(r, dict) and r.get("port") == port), None)
        if rule is not None:
            if rule.get("action") == "allow":
                allowed.append(port)
        elif rules.get("defaultPolicy") == "allow":
            allowed.append(port)
    return allowed


def _valid_ruleset(a: Any) -> bool:
    if not isinstance(a, dict) or a.get("defaultPolicy") not in ("allow", "deny"):
        return False
    rules = a.get("rules")
    if not isinstance(rules, list):
        return False
    return all(isinstance(r, dict) and is_int(r.get("port")) and r.get("action") in ("allow", "deny") for r in rules)


def check_answer(question: dict[str, Any], answer: Any) -> CheckResult:
    kind = question["kind"]

    if kind == "mc":
        return CheckResult(is_int(answer) and answer == question["answer"], [])

    if kind in ("match", "classify"):
        expected: list[int] = question["answer"]
        n = len(expected)
        if not _is_int_list(answer) or len(answer) != n:
            return CheckResult(False, [False] * n)
        items = [answer[i] == v for i, v in enumerate(expected)]
        return CheckResult(all(items), items)

    if kind == "numeric":
        expected_values: list[str] = question["answer"]
        n = len(expected_values)
        if not isinstance(answer, list) or len(answer) != n or not all(isinstance(x, str) for x in answer):
            return CheckResult(False, [False] * n)
        items = []
        for i, field in enumerate(question["fields"]):
            given = answer[i]
            want = expected_values[i]
            if field["format"] == "ipv4":
                items.append(normalize_ipv4(given) == want)
            elif field["format"] == "int":
                items.append(normalize_int(given) == want)
            else:
                items.append(normalize_text(given) == normalize_text(want))
        return CheckResult(all(items), items)

    if kind == "rules":
        ports = [p["port"] for p in question["ports"]]
        if not _valid_ruleset(answer):
            return CheckResult(False, [False] * len(ports))
        allowed = set(effective_allowed(ports, answer))
        wanted = set(question["answer"]["allowed"])
        items = [(p in allowed) == (p in wanted) for p in ports]
        return CheckResult(all(items), items)

    raise ValueError(f"tipo de questão desconhecido: {kind}")

File: packages/sim/src/terminals/check.ts
import type { AnswerValue, CheckResult, Question, RulesAnswer } from "./types";

/** Normaliza um IPv4 digitado ("192.168.001.010 " → "192.168.1.10"); retorna null se inválido. */
export function normalizeIpv4(s: string): string | null {
  const parts = s.trim().split(".");
  if (parts.length !== 4) return null;
  const nums: number[] = [];
  for (const p of parts) {
    if (!/^\d{1,3}$/.test(p)) return null;
    const n = Number(p);
    if (n > 255) return null;
    nums.push(n);
  }
  return nums.join(".");
}

export function normalizeInt(s: string): string | null {
  const t = s.trim().replace(/^\+/, "");
  if (!/^\d{1,10}$/.test(t)) return null;
  return String(Number(t));
}

export function normalizeText(s: string): string {
  return s.trim().toLowerCase().replace(/\s+/g, " ");
}

const isIntArray = (a: unknown): a is number[] => Array.isArray(a) && a.every((x) => Number.isInteger(x));

/** Portas liberadas por um conjunto de regras, avaliado apenas sobre as portas do cenário. */
export function effectiveAllowed(ports: readonly number[], rules: RulesAnswer): number[] {
  return ports.filter((port) => {
    const r = rules.rules.find((x) => x.port === port);
    return r ? r.action === "allow" : rules.defaultPolicy === "allow";
  });
}

export function checkAnswer(q: Question, a: AnswerValue | unknown): CheckResult {
  switch (q.kind) {
    case "mc":
      return { correct: Number.isInteger(a) && a === q.answer, items: [] };
    case "match":
    case "classify": {
      const n = q.answer.length;
      if (!isIntArray(a) || a.length !== n) return { correct: false, items: new Array<boolean>(n).fill(false) };
      const items = q.answer.map((v, i) => a[i] === v);
      return { correct: items.every(Boolean), items };
    }
    case "numeric": {
      const n = q.answer.length;
      if (!Array.isArray(a) || a.length !== n || !a.every((x) => typeof x === "string")) return { correct: false, items: new Array<boolean>(n).fill(false) };
      const items = q.fields.map((f, i) => {
        const given = a[i] as string;
        const expected = q.answer[i]!;
        if (f.format === "ipv4") return normalizeIpv4(given) === expected;
        if (f.format === "int") return normalizeInt(given) === expected;
        return normalizeText(given) === normalizeText(expected);
      });
      return { correct: items.every(Boolean), items };
    }
    case "rules": {
      const ports = q.ports.map((p) => p.port);
      const valid = !!a && typeof a === "object" && !Array.isArray(a)
        && ((a as RulesAnswer).defaultPolicy === "allow" || (a as RulesAnswer).defaultPolicy === "deny")
        && Array.isArray((a as RulesAnswer).rules)
        && (a as RulesAnswer).rules.every((r) => r && Number.isInteger(r.port) && (r.action === "allow" || r.action === "deny"));
      if (!valid) return { correct: false, items: ports.map(() => false) };
      const allowed = new Set(effectiveAllowed(ports, a as RulesAnswer));
      const expected = new Set(q.answer.allowed);
      const items = ports.map((p) => allowed.has(p) === expected.has(p));
      return { correct: items.every(Boolean), items };
    }
  }
}

/** Gabarito no formato em que o cliente envia a resposta (ganchos de teste e fixtures). */
export function canonicalAnswer(q: Question): AnswerValue {
  switch (q.kind) {
    case "mc":
      return q.answer;
    case "match":
    case "classify":
    case "numeric":
      return q.answer;
    case "rules":
      return { defaultPolicy: "deny", rules: q.answer.allowed.map((port) => ({ port, action: "allow" as const })) };
  }
}
